---
title: "What is an Exception?"
linkTitle: "What is an Exception?"
description: "Information regarding the anatomy of an exception, and properties accessible on an exception."
weight: 1
---

# {{% param title %}}

## Summary

An exception is thrown during an [execution][] when an issue is encountered whilst executing a [flow][]. Exceptions can occur for a number of reasons, such as:

- Exceptions thrown by a [block][], these can be found in the relevant documentation for that [block][All Blocks] (e.g. A duplicate key is added to a [Dictionary][] using the [Add Item With Key][] block)
- Developer thrown [exceptions][Throwing Exceptions] (e.g. A developer enriches an exception that has already been thrown to provide a more detailed message)
- Issues with third-party systems (e.g. A third-party system experiences an error)
- Issues with network connectivity (e.g. A target machine is inaccessible)
- Issues with permissions (e.g. The execution does not have permission to read a file)
- Issues with the environment (e.g. The software installed on a target machine is incorrect)

An exception contains details of what caused the issue, for more information see [Anatomy of an Exception][].

An exception can be handled in different ways depending on where the exception occurred. For more information see [Handling Exceptions][].

## Anatomy of an Exception

{{< figure src="/images/exception-in-viewer.PNG" >}}

Data within an exception can be accessed using properties.

The following properties are found on all exceptions:

| Property | Notes | Example |
|----------|-------|---------|
| Exception Type | The type of the exception | `PropertyNullException`
| Message | The exception message, providing information about the exception that occurred. This typically includes the cause of the exception, possible resolutions, and any related information. What is included in an exception's message can be found in the relevant documentation for that [exception][Exceptions] | `'List' is null; it must be provided with a non-null value.\r\nPlease click the HelpLink for more information on how to fix this.`
| HelpLink | Help link to aid with troubleshooting and resolution for the exception. There are three types of help links:</br></br><ul><li>Cortex - Help links direct to material within the Cortex Product Portal. This documentation is written and maintained by Cortex and contains information directly related to the Cortex product.</li><li>External - Help links direct to externally hosted material about the exception. This documentation is not written or maintained by Cortex.</li><li>Custom - Help links are generated by flow developers during the design and development of flows. These help links are configured in the [Throw New Exception][] block, and can link to any resource applicable to the exception.</li></ul> | `https://v2022.docs.cortex-ia.com/docs/reference/exceptions/common/property/property-null-exception/` |

The following properties are found on many exceptions:

| Property | Notes | Example |
|----------|-------|---------|
| Category | The category of the exception, which is used to categorise an exception if there are multiple reasons that the exception can occur. | [InfiniteLoopException][InfiniteLoopException: Category] has one category, `"Increment"` |
| Error Code | The error code for the exception, which is used to indicate the reason that the exception occurred, if there are multiple reasons that the exception can occur. | [InfiniteLoopException][InfiniteLoopException: Error Code] has three error codes, `InfiniteLoopErrorCode.IncrementIsZero (100)`, `InfiniteLoopErrorCode.IncrementIsNegative (101)`, and `InfiniteLoopErrorCode.IncrementIsPositive (102)` |
| Inner Exception | Contains the exception that caused this exception, if any. This property is always accessible but will only be visible in the [Exceptions Viewer][] when it contains data. | `{"ExceptionType": "PropertyNullException", "Message": "'List' is null; it must be provided with a non-null value.\r\nPlease click the HelpLink for more information on how to fix this.", "HelpLink": "https://v2022.docs.cortex-ia.com/docs/reference/exceptions/common/property/property-null-exception/"}` |

For more information regarding what properties are accessible on an exception, see the relevant documentation for that [exception][Exceptions].  

## Remarks

### Known Limitations

#### Carriage Return and Line Feed Characters

Carriage Return and Line Feed characters within an exception are supported but will be displayed or written as `\r`, `\n`, or `\r\n`.

In future this limitation may be removed.

#### Copying Exception Text

The [Exceptions Viewer][] currently does not support traditional copy functionality. A workaround is to highlight the text, then click and drag into another text editor.

In future this limitation may be removed.

## See Also

### Related Concepts

- [Flows][]
- [Blocks][]
- [Executions][]

### Related Data Types

- [All Exceptions][Exceptions]

### Related Blocks

- [All Blocks][]

### External Documentation

None

[Anatomy of an Exception]: {{< ref "#anatomy-of-an-exception" >}}

[Handling Exceptions]: {{< url path="Cortex.Reference.Concepts.Fundamentals.Exceptions.HandlingExceptions.MainDoc" >}}
[Throwing Exceptions]: {{< url path="Cortex.Reference.Concepts.Fundamentals.Exceptions.ThrowingExceptions.MainDoc" >}}

[Blocks]: {{< url path="Cortex.Reference.Concepts.Fundamentals.Blocks.MainDoc" >}}
[block]: {{< url path="Cortex.Reference.Concepts.Fundamentals.Blocks.WhatIsABlock.MainDoc" >}}

[Executions]: {{< url path="Cortex.Reference.Concepts.Fundamentals.Executions.MainDoc" >}}
[execution]: {{< url path="Cortex.Reference.Concepts.Fundamentals.Executions.WhatIsAnExecution.MainDoc" >}}

[Flows]: {{< url path="Cortex.Reference.Concepts.Fundamentals.Flows.MainDoc" >}}
[flow]: {{< url path="Cortex.Reference.Concepts.Fundamentals.Flows.WhatIsAFlow.MainDoc" >}}

[All Blocks]: {{< url path="Cortex.Reference.Blocks.MainDoc" >}}
[Exception Blocks]: {{< url path="Cortex.Reference.Blocks.Exceptions.MainDoc" >}}
[Throw New Exception]: {{< url path="Cortex.Reference.Blocks.Exceptions.ThrowException.ThrowNewFlowException.MainDoc" >}}
[Add Item With Key]: {{< url path="Cortex.Reference.Blocks.Dictionaries.AddItem.AddItemWithKey.MainDoc" >}}

[Dictionary]: {{< url path="Cortex.Reference.DataTypes.Collections.Dictionary.MainDoc" >}}

[Exceptions]: {{< url path="Cortex.Reference.Exceptions.MainDoc" >}}
[InfiniteLoopException: Category]: {{< url path="Cortex.Reference.Exceptions.Loops.InfiniteLoopException.Category" >}}
[InfiniteLoopException: Error Code]: {{< url path="Cortex.Reference.Exceptions.Loops.InfiniteLoopException.ErrorCode" >}}

[Exceptions Viewer]: {{< url path="Cortex.Guides.Studio.EastPanel.ExceptionsViewer" >}}
